Description: Support for GNU/Hurd
Author: Sergey B Kirpichev <skirpichev@gmail.com>
Forwarded: https://bitbucket.org/tildeslash/monit/pull-request/3

 1. Enable building for GNU/Hurd in configure.ac
 2. Fix MAX_PATH use cases
 3. Implement minimal src/{device,process}/sysdep_HURD.c
 4. Minor adaptation of sysdep_LINUX.c: revert _getStartTime() to pre-5.16
    version (function was get_starttime())

---
 configure.ac                         |    5 +++++
 libmonit/configure.ac                |    4 ++++
 libmonit/src/Config.h                |    6 ++++++
 libmonit/src/system/os/hurd/Link.inc |    1 +
 src/device/sysdep_HURD.c             |    1 +
 src/monit.h                          |    5 +++++
 src/process/sysdep_HURD.c            |   20 ++++++++++++++++++++
 src/process/sysdep_LINUX.c           |   11 ++++++++---
 8 files changed, 50 insertions(+), 3 deletions(-)

--- a/configure.ac
+++ b/configure.ac
@@ -583,6 +583,11 @@ then
    CFLAGS=`echo $CFLAGS|sed 's/-g//g'`
    CFLAGS="$CFLAGS -D_THREAD_SAFE -D_REENTRANT"
    LIBS="$LIBS -lodm -lperfstat -lm"
+elif test "$architecture" = "GNU"
+then
+   ARCH="HURD"
+   CFLAGS="$CFLAGS -D _REENTRANT"
+   LDFLAGS="$LDFLAGS -rdynamic"
 else
    AC_MSG_WARN([Architecture not supported: ${architecture}])
    CFLAGS="$CFLAGS -D _REENTRANT"
--- a/libmonit/configure.ac
+++ b/libmonit/configure.ac
@@ -285,6 +285,10 @@ then
    CFLAGS="$CFLAGS -D_THREAD_SAFE -D_REENTRANT"
    AC_DEFINE([AIX], 1, [Define to 1 if the system is AIX])
    LIBS="$LIBS -lcfg -lodm -lperfstat"
+elif test "$architecture" = "GNU"
+then
+   CFLAGS="$CFLAGS -D _REENTRANT"
+   AC_DEFINE([HURD], 1, [Define to 1 if the system is Hurd])
 else
    AC_MSG_ERROR([Architecture not supported: ${architecture}])
 fi
--- /dev/null
+++ b/src/device/sysdep_HURD.c
@@ -0,0 +1 @@
+#include "sysdep_LINUX.c"
--- /dev/null
+++ b/src/process/sysdep_HURD.c
@@ -0,0 +1,20 @@
+#include "config.h"
+
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+
+/**
+ * On GNU/Hurd procps usually launched with -c (linux-compatibility switch):
+ * http://lists.debian.org/debian-hurd/2012/10/msg00070.html
+ * so, sysconf(_SC_CLK_TCK_) is useless here right now...
+ */
+#ifndef HZ
+# ifndef HURD
+#  define HZ sysconf(_SC_CLK_TCK)
+# else
+#  define HZ 100
+# endif
+#endif
+
+#include "sysdep_LINUX.c"
--- a/libmonit/src/Config.h
+++ b/libmonit/src/Config.h
@@ -91,6 +91,12 @@
 typedef unsigned char uchar_t;
 #endif
 
+/**
+ * For systems without PATH_MAX define it with a resonable value
+ */
+#ifndef PATH_MAX
+#define PATH_MAX 1024
+#endif
 
 #endif
 
--- a/src/monit.h
+++ b/src/monit.h
@@ -417,6 +417,11 @@ Sigfunc *signal(int signo, Sigfunc * fun
 #define SIG_ERR ((Sigfunc *)-1)
 #endif
 
+/* For systems without PATH_MAX define it with a resonable value */
+
+#ifndef PATH_MAX
+#define PATH_MAX 1024
+#endif
 
 /** ------------------------------------------------- General purpose macros */
 
--- /dev/null
+++ b/libmonit/src/system/os/hurd/Link.inc
@@ -0,0 +1 @@
+#include "../linux/Link.inc"
--- a/src/process/sysdep_LINUX.c
+++ b/src/process/sysdep_LINUX.c
@@ -179,12 +179,17 @@ static double hz = 0.;
  * @return seconds since unix epoch
  */
 static time_t _getStartTime(void) {
-        struct sysinfo info;
-        if (sysinfo(&info) < 0) {
+        char buf[STRLEN];
+        double up = 0;
+        if (! file_readProc(buf, sizeof(buf), "uptime", -1, NULL)) {
                 Log_error("system statistic error -- cannot get system uptime: %s\n", STRERROR);
                 return 0;
         }
-        return Time_now() - info.uptime;
+        if (sscanf(buf, "%lf", &up) != 1) {
+                Log_error("system statistic error -- invalid uptime\n");
+                return 0;
+        }
+        return Time_now() - (time_t)up;
 }
 
 
